### PR TITLE
Rework tracing to exclude SMT setup times

### LINKED ISSUE #565: `/trace` prints incorrect time for first assertion in each file
Boogie's `/trace` seems to include some startup/warmup cost in the time of the first implementation in each file; for example, the following program…

```
procedure m1();
implementation m1() {
  assert true;
}

procedure m2();
implementation m2() {
  assert true;
}
```

… prints this:

```
Verifying m1 ...
  [0.149 s, 1 proof obligation]  verified

Verifying m2 ...
  [0.007 s, 1 proof obligation]  verified
```

This makes it hard to reliably find performance issues in a Boogie file.

### PR DESCRIPTION
**Pull request by cpitclaudel**
Closes #565.

This is not very polished but it seems to work for me.  I haven't had the time to test it with splits.

Output before:

```
Verifying m1 ...
[TRACE] Using prover: z3
  [0.345 s, solver resource count: 92, 1 proof obligation]  verified

Verifying m2 ...
  [0.062 s, solver resource count: 92, 1 proof obligation]  verified
```

Output after:

```
Verifying m1 ...
[TRACE] Using prover: z3
  [0.057 s, solver resource count: 92, 1 proof obligation]  verified

Verifying m2 ...
  [0.081 s, solver resource count: 92, 1 proof obligation]  verified
```

### CODE REVIEW

File: Source/VCGeneration/VCGen.cs
Comment (atomb) [inline Source/VCGeneration/VCGen.cs:362]: Is this `Stopwatch` still used anywhere after removing the code below?

**Review by atomb — DISMISSED** [review]
This looks good to me. I left one question, but I'm not sure whether it requires any changes.

File: Source/ExecutionEngine/VerificationResult.cs
Comment (shazqadeer) [inline Source/ExecutionEngine/VerificationResult.cs:24]: a the -> the

more narrow -> narrower

**Review by shazqadeer — DISMISSED** [review]

File: Source/ExecutionEngine/VerificationResult.cs
Comment (cpitclaudel) [inline Source/ExecutionEngine/VerificationResult.cs:24]:
```suggestion
  /// Gets or sets the elapsed time, excluding SMT solver setup costs.
```

File: Source/ExecutionEngine/VerificationResult.cs
Comment (cpitclaudel) [inline Source/ExecutionEngine/VerificationResult.cs:25]:
```suggestion
  /// (This is narrower than <c>Start - End</c>.)
```

File: Source/VCGeneration/VCGen.cs
Comment (cpitclaudel) [reply Source/VCGeneration/VCGen.cs:362]: Nope, it's a rebase gone wrong.  Thanks!

Comment (cpitclaudel) [inline Source/VCGeneration/VCGen.cs:362]:
```suggestion
```

**Comment by cpitclaudel** [comment]
Thanks all for the reviews!  I'd like to wait for a review from @keyboardDrummer , since he recently modified this bit of the code (and it would be bad if we gave incorrect times)

File: Source/VCGeneration/Checker.cs
Comment (keyboardDrummer) [inline Source/VCGeneration/Checker.cs:252]: 'How' seems superfluous

File: Source/VCGeneration/Checker.cs
Comment (keyboardDrummer) [inline Source/VCGeneration/Checker.cs:253]: What startup costs, starting the solver is already done before ProverStateTime, right?

**Comment by shazqadeer** [comment]
@cpitclaudel : It would be good to land this PR soon. I see that we already have some conflicts presumably from the recent PR landed by @atomb. Thanks.

**Comment by shazqadeer** [comment]
@cpitclaudel : Should we abandon this PR?

**Comment by shazqadeer** [comment]
Thank you. Much appreciated.

File: Source/VCGeneration/Checker.cs
Comment (cpitclaudel) [reply Source/VCGeneration/Checker.cs]: It was poor phrasing / a typo.  There's no `ProverStartTime`; it's `VerificationResult.Start` that I meant.

File: Source/Provers/SMTLib/SMTLibInteractiveTheoremProver.cs
Comment (cpitclaudel) [reply Source/Provers/SMTLib/SMTLibInteractiveTheoremProver.cs]: They are not (intended to be) different; it's just that using `end time - start time` to measure elapsed time is not great (because these clocks are not guaranteed to be monotonic and are not very precise anyway).  

However, there is no good reason to have them set in separate places. I cleaned things up by rewriting the code to start the stopwatch when `ProverStartTime` is set and consult it when the elapsed time was computed.

**Comment by cpitclaudel** [comment]
Updated and cleaned up.  I've simplified the timing code based on @keyboardDrummer feedback, and documented the remaining shortcomings.

**Review by keyboardDrummer — APPROVED** [review]

**Pull request merged**
